crates/lair_keystore_api/src/config.rs

```diff
@@ -0,0 +1,254 @@
+//! Lair Configuration Types
+
+use crate::prelude::*;
+use std::future::Future;
+use std::sync::Arc;
+
+/// Config used by lair servers.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct LairServerConfigInner {
+    /// The connection url for communications between server / client.
+    /// - `unix:///path/to/unix/socket?k=Yada`
+    /// - `named_pipe:\\.\pipe\my_pipe_name?k=Yada`
+    /// - `tcp://127.0.0.1:12345?k=Yada`
+    pub connection_url: url::Url,
+
+    /// The pid file for managing a running lair-keystore process
+    pub pid_file: std::path::PathBuf,
+
+    /// The sqlcipher store file for persisting secrets
+    pub store_file: std::path::PathBuf,
+
+    /// salt for decrypting runtime data
+    pub runtime_secrets_salt: BinDataSized<16>,
+
+    /// argon2id mem_limit for decrypting runtime data
+    pub runtime_secrets_mem_limit: u32,
+
+    /// argon2id ops_limit for decrypting runtime data
+    pub runtime_secrets_ops_limit: u32,
+
+    /// the runtime context key secret
+    pub runtime_secrets_context_key: SecretDataSized<32, 49>,
+
+    /// the server identity signature keypair seed
+    pub runtime_secrets_sign_seed: SecretDataSized<32, 49>,
+}
+
+impl std::fmt::Display for LairServerConfigInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = serde_yaml::to_string(&self).map_err(|_| std::fmt::Error)?;
+        f.write_str(&s)
+    }
+}
+
+impl LairServerConfigInner {
+    /// decode yaml bytes into a config struct
+    pub fn from_bytes(bytes: &[u8]) -> LairResult<Self> {
+        serde_yaml::from_slice(bytes).map_err(one_err::OneErr::new)
+    }
+
+    /// Construct a new default lair server config instance.
+    /// Respects hc_seed_bundle::PwHashLimits.
+    pub fn new<P>(
+        root_path: P,
+        passphrase: sodoken::BufRead,
+    ) -> impl Future<Output = LairResult<Self>> + 'static + Send
+    where
+        P: AsRef<std::path::Path>,
+    {
+        let root_path = root_path.as_ref().to_owned();
+        let limits = hc_seed_bundle::PwHashLimits::current();
+        async move {
+            // default pid_file name is '[root_path]/pid_file'
+            let mut pid_file = root_path.clone();
+            pid_file.push("pid_file");
+
+            // default store_file name is '[root_path]/store_file'
+            let mut store_file = root_path.clone();
+            store_file.push("store_file");
+
+            // generate a random salt for the pwhash
+            let salt = <sodoken::BufWriteSized<16>>::new_no_lock();
+            sodoken::random::bytes_buf(salt.clone()).await?;
```

Comment:
this clone confused me, i wasn't sure how the randomness was useful in a clone of `salt` then i saw the arc in `BufWriteSized` - maybe a comment for that?

Reply:
Yeah, I can add some documentation to the sodoken crate about this peculiarity in the api... I'm not sure I want to comment every usage of it downstream though : )